### PR TITLE
[Refactor] use `GenericMessage` in `createOrUpdate(...)`

### PR DESCRIPTION
**Pull request by David-Henner**
## What's new in this PR?

- Replaced `ZMGenericMessage` by `GenericMessage` in `ZMOTRMessage.createOrUpdate(...)` method
- Converted necessary protobuf objects to Swift protobufs
- Converted some objc methods to Swift to use Swift protobufs
- Updated tests

### CODE REVIEW

File: Source/Model/Message/ZMOTRMessage+UpdateEvent.swift [hunk 1]
```diff
@@ -42,37 +48,46 @@ extension ZMOTRMessage {
...
-        } else if message.hasButtonActionConfirmation() {
+        case .buttonActionConfirmation:
             ZMClientMessage.updateButtonStates(withConfirmation: message.buttonActionConfirmation, forConversation: conversation, inContext: moc)
-        } else if message.hasEdited() {
+        case .edited:
             return ZMClientMessage.editMessage(withEdit: message.edited, forConversation: conversation, updateEvent: updateEvent, inContext: moc, prefetchResult: prefetchResult)
-        } else if conversation.shouldAdd(event: updateEvent) && !(message.hasClientAction() || message.hasCalling() || message.hasAvailability()) {
-            guard let nonce = UUID(uuidString: message.messageId) else { return nil }
-            let messageClass: AnyClass = ZMGenericMessage.entityClass(for: message)
+        case .clientAction, .calling, .availability:
```
Comment (KaterinaWire) [inline Source/Model/Message/ZMOTRMessage+UpdateEvent.swift:80]: The original code contains` !(message.hasClientAction()`

File: Source/Utilis/Protos/GenericMessage+MessageCapable.swift [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        set {}
+    }
+    
+    var defaultLegalHoldStatus: LegalHoldStatus {
+        return .unknown
+    }
+}
+
+extension GenericMessage.OneOf_Content {
+    var messageCapable: MessageCapable? {
```
Comment (KaterinaWire) [inline Source/Utilis/Protos/GenericMessage+MessageCapable.swift:49]: Why do we need this extension?

File: Source/Model/Message/ZMOTRMessage+UpdateEvent.swift (hunk 1 above)
Comment (David-Henner) [reply Source/Model/Message/ZMOTRMessage+UpdateEvent.swift:80]: I don't see the issue ?

File: Source/Utilis/Protos/GenericMessage+MessageCapable.swift (hunk 2 above)
Comment (KaterinaWire) [inline Source/Utilis/Protos/GenericMessage+MessageCapable.swift:49]: Why do we need this extension?

Comment (David-Henner) [reply Source/Utilis/Protos/GenericMessage+MessageCapable.swift:49]: it's used in `ZMConversation+SecurityLevel` in the func `updateLegalHoldIfNeededWithHint`

Comment (David-Henner) [reply Source/Utilis/Protos/GenericMessage+MessageCapable.swift:49]: we need it to access the `legalHoldStatus`

File: Source/Model/Message/ZMOTRMessage+UpdateEvent.swift (hunk 1 above)
Comment (KaterinaWire) [reply Source/Model/Message/ZMOTRMessage+UpdateEvent.swift:80]: According to the original code, we should not return nil if message.hasCalling() || message.hasAvailability().

File: Source/Utilis/Protos/GenericMessage+MessageCapable.swift (hunk 2 above)
Comment (KaterinaWire) [reply Source/Utilis/Protos/GenericMessage+MessageCapable.swift:49]: Maybe we can try like this:

```
extension GenericMessage {
    var legalHoldStatus: LegalHoldStatus? {
        guard let content = content else {
            return nil
        }
        
        switch content {
        case .ephemeral:
            return ephemeral.legalHoldStatus
        case .reaction:
            return reaction.legalHoldStatus
        case .knock:
            return knock.legalHoldStatus
        case .text:
            return text.legalHoldStatus
        case .location:
            return location.legalHoldStatus
        case .asset:
            return asset.legalHoldStatus
        default:
            return nil
        }
    }
}
```

What do you think?

File: Source/Model/Message/ZMOTRMessage+UpdateEvent.swift (hunk 1 above)
Comment (David-Henner) [reply Source/Model/Message/ZMOTRMessage+UpdateEvent.swift:80]: let's take a look a the original code

```
...
else if conversation.shouldAdd(event: updateEvent) && !(message.hasClientAction() || message.hasCalling() || message.hasAvailability()) {
   ...
   return clientMessage
}
return nil
```

For the sake of the conversation, let's assume `conversation.shouldAdd(event: updateEvent)` is always true
We end up with the following statement
```
else if !(message.hasClientAction() || message.hasCalling() || message.hasAvailability())
```

Let's say `message.hasClientAction()` returns `true` and the rest is `false`

```
else if !(true || false || false)

// or 

else if !true
```

In this case, the condition evaluates to `false` and we `return nil`
We can then say that if any of the conditions (`message.has...`) are true, we should return nil
That can translate to `case .clientAction, .calling, .availability: return nil`

File: Source/Utilis/Protos/GenericMessage+MessageCapable.swift (hunk 2 above)
Comment (David-Henner) [reply Source/Utilis/Protos/GenericMessage+MessageCapable.swift]: Yep, much better.
I updated it in ee2cb4a.
It seems that we don't need `legalHoldStatus` in `MessageCapable` protocol, so I removed it for simplicity

**Pull request merged**